File: app/src/main/java/uk/co/sundroid/activity/data/fragments/dialogs/date/YearPickerFragment.java
package uk.co.sundroid.activity.data.fragments.dialogs.date;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.NumberPicker;
import uk.co.sundroid.R.id;
import uk.co.sundroid.R.layout;

import java.util.Calendar;

import static java.util.Calendar.YEAR;
import static java.util.Calendar.getInstance;

public class YearPickerFragment extends DialogFragment implements DialogInterface.OnClickListener {

    private int year;
    private NumberPicker yearPicker;

    @FunctionalInterface
    public interface OnYearSelectedListener {
        void onYearSet(int year);
    }

    public static YearPickerFragment newInstance(Calendar calendar) {
        YearPickerFragment fragment = new YearPickerFragment();
        Bundle args = new Bundle();
        args.putInt("y", calendar.get(YEAR));
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Calendar today = getInstance();
        this.year = getArguments().getInt("y", today.get(YEAR));
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        setRetainInstance(true); // TODO May not be required
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        View content = View.inflate(getActivity(), layout.dialog_yearpicker, null);
        yearPicker = content.findViewById(id.yearPicker);
        yearPicker.setMinValue(2000);
        yearPicker.setMaxValue(2020);
        yearPicker.setWrapSelectorWheel(false);
        yearPicker.setValue(year);
        builder.setView(content);
        builder.setTitle("Set year");
        builder.setPositiveButton("Set", this);
        builder.setNeutralButton("This year", this);
        builder.setNegativeButton("Cancel", this);
        return builder.create();
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int button) {
        Fragment target = getTargetFragment();
        if (target != null && target instanceof OnYearSelectedListener) {
            if (button == DialogInterface.BUTTON_NEUTRAL) {
                Calendar today = getInstance();
                ((OnYearSelectedListener)target).onYearSet(today.get(YEAR));
            } else if (button == DialogInterface.BUTTON_POSITIVE) {
                ((OnYearSelectedListener)target).onYearSet(yearPicker.getValue());
            }
        }
        dismiss();
    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/dialogs/date/YearPickerFragment.kt
package uk.co.sundroid.activity.data.fragments.dialogs.date

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.view.View
import android.widget.NumberPicker
import androidx.fragment.app.DialogFragment
import uk.co.sundroid.R
import uk.co.sundroid.R.layout
import java.util.*
import java.util.Calendar.*

class YearPickerFragment : DialogFragment() {

    private var calendar: Calendar = getInstance()

    @FunctionalInterface
    interface OnYearSelectedListener {
        fun onYearSet(year: Int)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        restore(arguments)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        save(calendar, outState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        restore(savedInstanceState)

        val view = View.inflate(activity, layout.dialog_yearpicker, null)

        view.findViewById<NumberPicker>(R.id.yearPicker)?.apply {
            minValue = 2000
            maxValue = 2025
            wrapSelectorWheel = false
            value = calendar.get(YEAR)
            setOnValueChangedListener { _, _, year -> calendar.set(YEAR, year) }
        }

        return AlertDialog.Builder(activity).apply {
            setView(view)
            setTitle("Set year")
            setPositiveButton("Set") { _, _ -> set(calendar) }
            setNeutralButton("This year") { _, _ -> set(getInstance()) }
            setNegativeButton("Cancel", null)
        }.create()
    }

    private fun set(calendar: Calendar) {
        val target = targetFragment
        if (target is OnYearSelectedListener) {
            (target as OnYearSelectedListener).onYearSet(calendar.get(YEAR))
        }
    }

    private fun restore(bundle: Bundle?) {
        if (bundle != null) {
            val ymd = bundle.getIntArray("ymd")
            val tz = bundle.getString("tz")
            if (ymd != null && tz != null) {
                calendar.timeZone = TimeZone.getTimeZone(tz)
                calendar.set(ymd[0], ymd[1], ymd[2])
            }
        }
    }

    companion object {
        fun newInstance(calendar: Calendar): YearPickerFragment {
            return YearPickerFragment().apply {
                arguments = save(calendar)
            }
        }

        private fun save(calendar: Calendar, bundle: Bundle? = Bundle()): Bundle? {
            return bundle?.apply {
                putIntArray("ymd", intArrayOf(calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH)))
                putString("tz", calendar.timeZone.id)
            }
        }
    }

}